005-1-locations-on-stack.md

```diff
@@ -0,0 +1,590 @@
+>     variable a particular location refers to.
+>
+> 3. `DW_OP_bit_piece`
+>
+>     The DW_OP_bit_piece operation takes two operands. The first is an
+>     unsigned LEB128 number that gives the size `S`, in bits, of the piece.
+>     The second is an unsigned LEB128 number that gives the offset in bits
+>     from the location defined by the location `A` on the top of the stack.
+>
+>     Interpretation of the offset depends on the type of location. If the
+>     location is an undefined location (see Section 2.5.2.4), the
+>     `DW_OP_bit_piece` operation describes a piece consisting of the given
+>     number of bits whose values are undefined, and the offset is ignored. If
+>     the location is a memory address (see Section 2.5.2.1), the
+>     `DW_OP_bit_piece` operation describes a sequence of bits relative to the
+>     location whose address is on the top of the DWARF stack using the bit
+>     numbering and direction conventions that are appropriate to the current
+>     language on the target system. In all other cases, the source of the
+>     piece is given by either a register location (see Section 2.5.2.2) or an
+>     implicit value location (see Section 2.5.2.3); the offset is from the
+>     least significant bit of the source value.
+> 
+>     _`DW_OP_bit_piece` is used instead of `DW_OP_piece` when the piece to be
+>     assembled into a value or assigned to is not byte-sized or is not at the
+>     start of a register or addressable unit of memory._
+>
+>     _Whether or not a `DW_OP_piece` operation is equivalent to any
+>     `DW_OP_bit_piece` operation with an offset of 0 is ABI dependent._
+>
+> For compatibility with DWARF Version 5 and earlier, the following
+> additional rules apply to piece operations:
+>
+> - If a piece operation is processed while the stack is empty, a new
+> empty composite and an undefined location are pushed implicitly (as if
+> `DW_OP_composite DW_OP_undefined` had been processed immediately prior
+> to the piece operation). The result is a composite with a single
+> undefined piece.
+>
+> - Otherwise, if the top of the stack `A` is a composite, and is the only
+> element on the stack (i.e., `B` does not exist), an undefined location
+> is pushed implicitly (as if `DW_OP_undefined` had been processed
+> immediately prior to the piece operation), whereupon the composite `A`
+> becomes `B` and the undefined location is now `A`. The result is the
+> addition of an undefined piece to the existing composite location.
+>
+> - Otherwise, if the top of the stack `A` is a location, or convertible
+> to a location, and the preceding element is not a composite location,
+> one or more elements below `A` are popped and discarded until the
+> preceding element `B` is a composite location, or until `A` is the
+> only element on the stack. If `A` is the only remaining element, a new
+> empty composite is inserted before it (as if `DW_OP_composite
```

Comment:
> as if `DW_OP_composite DW_OP_swap` had been processed immediately prior to the piece operation

I understand what you mean, but that might be confusing, because if `DW_OP_composite DW_OP_swap` had occurred prior to the piece operation, then we wouldn't do all this popping and special case handling.